src/handle_commit.cxx

```diff
@@ -326,8 +326,14 @@ void raft_server::commit_app_log(ulong idx_to_commit,
                 switch (ctx_->get_params()->return_method_) {
                 case raft_params::blocking:
                 default:
-                    // Blocking mode: invoke waiting function.
-                    elem->awaiter_.invoke();
+                    // Blocking mode
+                    if (elem->timer_.get_ms() > ctx_->get_params()->client_req_timeout_) {
+                        // If client timeout, remove elem here
+                        commit_ret_elems_.erase(entry);
+                    } else {
+                        // Else notify client that request done
+                        elem->awaiter_.invoke();
+                    }
```

Comment:
Relying on a timer doesn't seem to be a good practice due to its non-deterministic value. I'd rather add a new boolean field to `commit_ret_elem`:
```c++
struct commit_ret_elem{
  ...
  bool callback_invoked_; // default = false
};
```
and inside `handle_cli_req_callback`, 
```c++
    {   auto_lock(commit_ret_elems_lock_);
        idx = elem->idx_;
        ...
        elem->callback_invoked_ = true;
    }
```
then we can make this if-clause:
```c++
if (elem->callback_invoked_) {
    commit_ret_elems_.erase(entry);
} else {
    elem->awaiter_.invoke();
}
```

Reply:
I also considered timer may not a good way, thanks